src/Microsoft.OData.Client/Microsoft.OData.Client.csproj

```diff
@@ -147,6 +147,7 @@
     </PackageReference>
     <PackageReference Include="System.ComponentModel" Version="4.3.0" />
     <PackageReference Include="System.ComponentModel.Annotations" Version="4.7.0" />
+    <PackageReference Include="System.Reflection.Emit" Version="4.7.0" />

```

Comment:
>System.Reflection.Emit [](start = 31, length = 22)

Is there an alternative solution without adding this new dependency?

Reply:
@xuzhg Here is the challenge I was trying to solve...

Consider the following reduced version of the `Sale` and `Product` models
```csharp
public class Sale
{
    public int Id { get; set; }
    public string ProductId { get; set; }
    public Product Product { get; set; }
    public decimal Amount { get; set; }
}

public class Product
{
    public string Id { get; set; }
    public string Name { get; set; }
}
```
Grouping by property `Name` of `Product` navigation property and calculating average `Amount`,
```csharp
dataServiceContext.Sales.GroupBy(d1 = d1.Product.Name, 
    (d1, d2) => new { ProductName = d1, AverageAmount = d2.Average(d3 => d3.Amount) });
```
that is translated to the following `Uri`,
```
http://ServiceRoot/Sales?$apply=groupby((Product/Name),aggregate(Amount with average as AverageAmount))
```
Executed against the service, the response would look as follows:
```json
{
    "@odata.context": "http://service/$metadata#Sales
    (Product(Name),AverageAmount)",
    "value": [
        {
            "@odata.id": null,
            "AverageAmount": 2.00,
            "Product": { "@odata.id": null, "Name": "P1" }
        },
        {
            "@odata.id": null,
            "AverageAmount": 3.00,
            "Product": { "@odata.id": null, "Name": "P2" }
        }
    ]
}
```
_Notice the nested structure representing the grouping property `Product/Name`_.

The challenge I faced is that our **materialization logic expects a type with a structure mapping the response**. For the above case, the type would be expected to have the following structure:
```csharp
public class TYPENAME
{
    public decimal AverageAmount { get; set; }
    public Product Product { get; set; }
}
```
The `Sale` type does not contain `AverageAmount` property and **its also not an open type**.
The anonymous type generated in the `GroupBy` expression does not contain `Product` property and it presents a challenge since the properties are `readonly`, i.e. no setters.
Furthermore, without going into greater details, creating a projection plan where the anonymous type constructor can be invoked with values extracted directly from the materializer presented a different challenge altogether.

I resorted to this workaround where I create a collectible (`RunAndCollect`) assembly containing a type with the expected structure. With this type, the existing projection plan compiler is able to compose a projection plan to be used for materialization.
I have spent quite a bit of time pursuing different alternatives without much success. I'd be welcome to any new ideas/help on how differently I could approach the challenge.

Reply:
I revised the approach and got rid of this logic for dynamically emitting a type in a collectible assemblies. The new approach is centred around the existing projection plan compiler used for `Select` projections